actors/miner/src/lib.rs

```diff
@@ -3580,6 +3840,311 @@ pub struct ReplicaUpdateInner {
+                decl.deadline,
+                decl.partition,
+                e
+            ));
+        }
+
+        for sc in &decl.sectors_with_claims {
+            let claims = get_claims(rt, &sc.maintain_claims)
+                .context(format!("failed to get claims for sector {}", sc.sector_number))?;
+
+            for claim in claims {
+                // check provider and sector matches
+                if claim.provider != rt.message().receiver().id().unwrap() {
+                    return Err(actor_error!(illegal_argument, "invalid declaration indexing wrong claim: expected provider {} but found {} ", rt.message().receiver().id().unwrap(), claim.provider));
+                }
+                if claim.sector != sc.sector_number {
+                    return Err(actor_error!(illegal_argument, "invalid declaration indexing wrong claim: expected sector {} but found {} ", sc.sector_number, claim.sector));
+                }
+
+                // check expiration does not exceed term max
+                if decl.new_expiration > claim.term_start + claim.term_max {
+                    return Err(actor_error!(forbidden, "new expiration {} declared for a sector containing a claim only allowing extension to {}", decl.new_expiration, claim.term_start + claim.term_max));
+                }
+
+                claim_space_by_sector
+                    .entry(sc.sector_number)
+                    .and_modify(|size| *size += claim.size.0)
+                    .or_insert(claim.size.0);
+            }
+        }
+    }
+    Ok(ExtendExpirationsInner {
+        extensions: extensions.into_iter().map(|e2| e2.into()).collect(),
+        claims: Some(claim_space_by_sector),
+    })
+}
+
+fn extend_sector_committment(
+    policy: &Policy,
+    curr_epoch: ChainEpoch,
+    new_expiration: ChainEpoch,
+    sector: &SectorOnChainInfo,
+    claim_space_by_sector: &BTreeMap<SectorNumber, u64>,
+) -> Result<SectorOnChainInfo, ActorError> {
+    if !can_extend_seal_proof_type(sector.seal_proof) {
+        return Err(actor_error!(
+            forbidden,
+            "cannot extend expiration for sector {} with unsupported \
+            seal type {:?}",
+            sector.sector_number,
+            sector.seal_proof
+        ));
+    }
+
+    // This can happen if the sector should have already expired, but hasn't
+    // because the end of its deadline hasn't passed yet.
+    if sector.expiration < curr_epoch {
+        return Err(actor_error!(
+            forbidden,
+            "cannot extend expiration for expired sector {} at {}",
+            sector.sector_number,
+            sector.expiration
+        ));
+    }
+
+    if new_expiration < sector.expiration {
+        return Err(actor_error!(
+            illegal_argument,
+            "cannot reduce sector {} expiration to {} from {}",
+            sector.sector_number,
+            new_expiration,
+            sector.expiration
+        ));
+    }
```

Comment:
Factor this out for re-use in the legacy method, `validate_extended_expiration` or similar